### PR TITLE
liftSH hack to solve memory issue

### PR DESCRIPTION
**Pull request by erivas**
Commit related to issue #322 changing liftSH.

### CODE REVIEW

File: lib/Echidna/Transaction.hs [hunk 1]
```diff
@@ -253,7 +252,13 @@ shrinkTx tx'@(Tx c _ _ _ gp (C _ v) (C _ t, C _ b)) = let
 -- | Lift an action in the context of a component of some 'MonadState' to an action in the
 -- 'MonadState' itself.
 liftSH :: (MonadState a m, Has b a) => State b x -> m x
-liftSH = S.state . runState . zoom hasLens
+liftSH = stateST . runState . zoom hasLens
+  -- This is the default state function written in terms of get and set:
+  where stateST :: MonadState s m => (s -> (a, s)) -> m a
+        stateST f = do s <- get
```
Comment (incertia) [inline lib/Echidna/Transaction.hs:258]: put this on a newline and indent 2 spaces

**Comment by incertia** [comment]
is there any reason why copying the mtl `state` function fixes this? is it due to not sufficient inlining or something?

File: lib/Echidna/Transaction.hs (hunk 1 above)
Comment (erivas) [reply lib/Echidna/Transaction.hs:258]: Sorry, I don't get it. Should the `where` be in the same line as the `liftSH` and the `stateST` start in a newline with indent 2?
```
liftSH = stateST . runState . zoom hasLens where
  -- This is the default state function written in terms of get and set:
  stateST :: MonadState s m => (s -> (a, s)) -> m a
  stateST f = do s <- get
```

**Comment by erivas** [comment]
> is there any reason why copying the mtl `state` function fixes this? is it due to not sufficient inlining or something?

My guess was that the particular instance of `m` used is overriding the default implementation of `state` (which is part of the class `MonadState`)?

**Comment by ggrieco-tob** [comment]
According to [your research](https://github.com/crytic/echidna/issues/322#issuecomment-612932417), this PR also needs to limit the amount of gas per transaction to be effective. Please limit the amount of gas to the block limit (`8000030`) in the following lines:
* The gas used during the contract deployment ([here](https://github.com/crytic/echidna/blob/master/lib/Echidna/Solidity.hs#L159) and [here](https://github.com/crytic/echidna/blob/master/lib/Echidna/Solidity.hs#L231)).
* The gas used during the transaction generation ([here](https://github.com/crytic/echidna/blob/master/lib/Echidna/Config.hs#L110)).

**Comment by ggrieco-tob** [comment]
It seems that one of the @agroce gas tests is failing, perhaps he can join the discussion here..

**Comment by ggrieco-tob** [comment]
I can confirm this is a good workaround to keep control of the memory used. Thanks a lot @erivas !

**Comment by agroce** [comment]
Not sure I follow what's going on with my code here?

**Comment by ggrieco-tob** [comment]
@erivas had to add [this](https://github.com/crytic/echidna/pull/423/files#diff-09634d206772efe68f447bd8802bdd10R2) in order to fix [the failing gas usage test](https://github.com/crytic/echidna/runs/585706667#step:13:106).

File: lib/Echidna/Transaction.hs (hunk 1 above)
Comment (incertia) [reply lib/Echidna/Transaction.hs:258]:
```haskell
where stateST f = do
  s <- get
  -- ...
```

**Pull request merged**